### DISCUSSION TITLE
Using kintone UI Component v1 in React

### DISCUSSION THREAD

**Discussion by dong0** [kintone-labs/kintone-ui-component#990] [Show and tell]
# Background

Some developers want to use kintone UI Component (KUC) v1 in their React project.
But now, we don't provide React version and don't have a plan to support it officially in the future.

Internally we are using Web Components so that developers can access it from React.
https://reactjs.org/docs/web-components.html#gatsby-focus-wrapper
So as an option for React users, we want to share the possible coding way with it.
Here is an unofficial guide on how to use KUC v1 in React.

**Unofficial** means that we don't test the behavior and support it.

# Note for Coding

We are not supporting React coding way officially now, so please develop on your own responsibility.

## React 19 Compatibility Update

We have tested our guide "Using kintone UI Component v1 in React" with the newly released React 19 and can confirm that the integration approaches described in this guide remain compatible with React 19, with one important note:

**Important API Change**: For React 18 and later (including React 19), the `ReactDOM.render()` method used in our original examples has been deprecated. Instead, you should use the `createRoot()` API. We have updated the relevant code examples in this guide to reflect this change.

Other than this API update, all the integration patterns described in this guide continue to work with React 19 without requiring additional modifications. 

## Alternative Integration Approach

The approach described in this guide uses direct DOM manipulation and refs to interact with KUC components.  However, this method only works in JavaScript environments.

For TypeScript projects, we have developed an alternative approach using the [@lit/react](https://lit.dev/docs/frameworks/react/) library. You can find this guide here: [Using kintone UI Component v1 in React with @lit/react](https://github.com/kintone-labs/kintone-ui-component/discussions/1608). This method creates proper React wrappers for KUC components with cleaner component interfaces and better TypeScript support.

# Guide

Developer needs to write in different way depending on the property type, event and method of KUC.

We created some sample codes for property type, event and method.
These samples are authored in "Hooks API" styles.
You can refer to them.

### String or Number type property

- These properties can be defined like a React attribute.
For example, "label" of Text component and "text" of Button component, they can work as expected.

- Pay attention to the name of property and its reflecting attribute.
property is "className", attribute is "class"
property is "visible", attribute is "hidden"
Note: only the property with the setting "reflect: true" have the reflecting attribute, not all.
So please be careful if you use setAttribute and getAttribute.

#### Sample code

- Setting the static value to "text" and "type" of Button component.
```jsx
<kuc-button
  text="Show text value"
  type="submit"
/>
```

- Setting the dynamic value to "label" and "error" of Text component.
```jsx
<kuc-text
  label={label}
  class="kuc-text-ui"
  error={error}
/>
```

- Setting a number to "duration" of Notification component.
```jsx
<kuc-notification
  duration={1000}
/>
```

### Boolean or Array type property

- These properties **cannot** be defined like a React attribute.
For example, "requiredIcon", "items" of Dropdown component.
We need to use a [ref](https://reactjs.org/docs/refs-and-the-dom.html) to set the value of them.

- You can use React hooks [useEffect](https://reactjs.org/docs/hooks-reference.html#useeffect) and [useRef](https://reactjs.org/docs/hooks-reference.html#useref) to implement it.

#### Sample code

- We set "requiredIcon" and "items" of Dropdown component in useEffect function.
kucDropdownRef is the ref of kuc-dropdown.
- On UI, when user changes the selected item, the old value and new value will show on page. 
```jsx
import "kintone-ui-component";
import React, { useEffect, useRef, useState } from "react";

export const KucDropdownSample = () => {
  const [detail, setDetail] = useState({ value: "", oldValue: "" });
  const kucDropdownRef = useRef(null);

  useEffect(() => {
    const dropdown = kucDropdownRef.current;
    if (dropdown) {
      dropdown.addEventListener("change", (event) => {
        console.log(event.detail);
        setDetail({
          value: event.detail.value,
          oldValue: event.detail.oldValue,
        });
      });

      dropdown.requiredIcon = true;
      dropdown.items = [
        {
          label: "------",
          value: "",
        },
        {
          label: "orange",
          value: "Orange",
        },
        {
          label: "apple",
          value: "Apple",
        },
        {
          label: "banana",
          value: "Banana",
        },
      ];
    }
  }, [kucDropdownRef]);

  return (
    <div style={{ display: "inline-flex" }}>
      <kuc-dropdown ref={kucDropdownRef} label="kuc dropdown" />
      <div>
        <div style={{ padding: "10px 0 0 10px" }}>Value: {detail.value}</div>
        <div style={{ padding: "10px 0 0 10px" }}>
          Old Value: {detail.oldValue}
        </div>
      </div>
    </div>
  );
};
```

### HTMLElement type property

- We need to use a [ref](https://reactjs.org/docs/refs-and-the-dom.html) to set the value of this type property.
For example, "content", "footer" of Dialog component

- About how to create a HTMLElement in React, you can refer to the following steps.
(1) Creating a HTMLElement by document.createElement().
(2) For React 18 and later (including React 19): Using createRoot() and root.render() to render a React element into this HTMLElement.
Then you can set this HTMLElement as "content" or "footer".

#### Sample code

- Defining two functions to create HTMLElement for setting "content" and "footer".
- Setting "content" and "footer" of Dialog component in useEffect function.
- On UI, user can click a button to open the dialog.
In the dialog, there are two buttons to change the Title of dialog dynamically.
```jsx
import "kintone-ui-component";
import React, { useEffect, useRef, useState } from "react";
import { createRoot } from "react-dom/client"; // Updated import for React 18+

export const KucDialogSample = () => {
  const [title, setTitle] = useState(new Date().toLocaleString());
  const kucDialogRef = useRef(null);
  const kucTextRef = useRef(null);

  const createContent = () => {
    const div = document.createElement("div");
    const root = createRoot(div); // Using createRoot for React 18+
    root.render(
      <React.Fragment>
        <kuc-text ref={kucTextRef} />
        <kuc-button
          text="Change Title"
          onClick={() => {
            setTitle(kucTextRef.current.value);
          }}
        />
      </React.Fragment>,
      div
    );
    return div;
  };

  const createFooter = () => {
    const div = document.createElement("div");
    const root = createRoot(div); // Using createRoot for React 18+
    root.render(
      <kuc-button
        text="Click Me"
        onClick={() => {
          setTitle(new Date().toLocaleString());
        }}
      />,
      div
    );
    return div;
  };

  useEffect(() => {
    const dialog = kucDialogRef.current;
    if (dialog) {
      dialog.addEventListener("close", (event) => {
        console.log(event);
      });

      dialog.content = createContent();
      dialog.footer = createFooter();
    }
  }, [kucDialogRef]);

  return (
    <div>
      <kuc-dialog ref={kucDialogRef} title={title} />
      <kuc-button
        text="Show Dialog"
        onClick={() => {
          kucDialogRef.current.open();
        }}
      />
    </div>
  );
};
```

## KUC component event & method

- We need to use a [ref](https://reactjs.org/docs/refs-and-the-dom.html) to call the event or method of KUC.<br>
Event: using "addEventListener"
e.g. kucNotificationRef.current.addEventListener("close", (event) => {});<br>
Method: calling them directly
e.g. kucNotificationRef.current.open();

#### Sample code

- Setting the "click" event function of button to open the Notification component.
"kucNotificationRef.current.open()" is called in "click" event function, it's a method of Notification component.
- Using addEventListener to add "close" event listener of Notification component in useEffect function.
"close" is an event of Notification component.
```jsx
import "kintone-ui-component";
import React, { useEffect, useRef } from "react";

export const KucNotificationSample = () => {
  const kucNotificationRef = useRef(null);

  useEffect(() => {
    const notification = kucNotificationRef.current;
    if (notification) {
      notification.addEventListener("close", (event) => {
        console.log(event);
      });
    }
  }, [kucNotificationRef]);

  return (
    <div>
      <kuc-notification
        ref={kucNotificationRef}
        text="Notification"
        type="info"
      />
      <kuc-button
        text="Open Notification"
        onClick={() => {
          kucNotificationRef.current.open();
        }}
      />
      <kuc-button
        text="Close Notification"
        onClick={() => {
          kucNotificationRef.current.close();
        }}
      />
    </div>
  );
};
```

**Comment by dong0** [comment]
## Overview

As we wrote in the main text, there is a version conflict issue when defining multiple kintone UI Component (KUC) packages on custom JS/plug-in at the same time.

#### When importing multiple KUC packages (ESM) and using custom HTML tags directly:

For example, we have two scripts that are packaged by webpack as follows.
Here, we have the v1.2.0 script loaded first and v1.3.1 loaded last.
- One script uses v1.2.0 and specifies the text color of Dropdown to be green.
- One script uses v1.3.1 and specifies the text color of Dropdown to be red.

We can see that the text color of Dropdown in the v1.3.1 script is overridden by v1.2.0 script.
![Custom element is overridden by the first loaded file](https://user-images.githubusercontent.com/3798705/182608680-c0e4a5b1-fe83-4ccf-918f-9e814119f107.png)

You can refer to the [Version conflicts issue and solution](https://kintone-ui-component.netlify.app/docs/guides/version-conflicts-issue-solution) for the issue detail.

#### This issue is solved in KUC v1.4.0.

From v1.4.0, you need to write with version number included tag like `<kuc-button-x-x-x />` and specify your expected version.
For example, `<kuc-button-1-4-0 />`, `<kuc-text-1-4-0 />` and so on.

In addition, class names have also been changed to avoid CSS style conflicts.
For example, `kuc-button-1-4-0__button kuc-button-1-4-0__button--normal`.
If you are customizing CSS styles with KUC class names, please be aware of this change.

There is some sample code for reference.

## Sample code

Assuming you are currently using KUC v1.4.0 in your project, you can refer to the following sample code.

- Text component
Setting the dynamic value to "label" and "error" of Text component.
```jsx
<kuc-text-1-4-0
  label={label}
  class="kuc-text-ui"
  error={error}
/>
```

- Notification and Button component
Using two buttons to open/close notification.
```jsx
import "kintone-ui-component";
import React, { useEffect, useRef } from "react";

export const KucNotificationSample = () => {
  const kucNotificationRef = useRef(null);

  useEffect(() => {
    const notification = kucNotificationRef.current;
    if (notification) {
      notification.addEventListener("close", (event) => {
        console.log(event);
      });
    }
  }, [kucNotificationRef]);

  return (
    <div>
      <kuc-notification-1-4-0
        ref={kucNotificationRef}
        text="Notification"
        type="info"
      />
      <kuc-button-1-4-0
        text="Open Notification"
        onClick={() => {
          kucNotificationRef.current.open();
        }}
      />
      <kuc-button-1-4-0
        text="Close Notification"
        onClick={() => {
          kucNotificationRef.current.close();
        }}
      />
    </div>
  );
};
```

In order to avoid modifying too much code after future upgrades, you can use the version function to generate the suffix of the tag.
Please refer to the example below.

- Button and Text component
```jsx
import React, { useEffect, useRef, useState } from "react";
import { version } from "kintone-ui-component";

export const KucTextVersionSample = () => {
  const [label, setLabel] = useState("");
  const [value, setValue] = useState("");

  const kucTextRef = useRef(null);

  useEffect(() => {
    const text = kucTextRef.current;
    if (text) {
      text.addEventListener("change", (event) => {
        setValue(event.detail.value);
      });
      text.requiredIcon = true;
    }
  }, [kucTextRef]);

  const versionSuffix = `-${version.replace(/\./g, "-")}`;
  const KucButton = `kuc-button${versionSuffix}`;
  const KucText = `kuc-text${versionSuffix}`;
  return (
    <div style={{ display: "flex" }}>
      {version}
      <KucButton
        text="Show text value"
        onClick={() => {
          setLabel(value);
        }}
        type="submit"
      />
      <KucText ref={kucTextRef} label={label} />
    </div>
  );
};
```

**Comment by memberwebm** [comment]
how to use kuc-table-1-14-0 in config, it show error: Error: 'columns' property is not array, could you help me?

**Reply by dong0** [reply]
I replied to you in the issue comment. https://github.com/kintone-labs/kintone-ui-component/issues/1484#issuecomment-1863925347
Please refer to it.
Thank you!
